service/controller/resource/clusterownerreference/resource.go

```diff
@@ -86,16 +118,79 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	}
 
 	r.logger.LogCtx(ctx, "message", fmt.Sprintf("Ensured %s label and 'ownerReference' fields on AzureCluster '%s/%s'", capiv1alpha3.ClusterLabelName, cluster.Namespace, cluster.Spec.InfrastructureRef.Name))
-
 	return nil
 }
 
-// EnsureDeleted is no-op.
-func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+func (r *Resource) ensureMachinePools(ctx context.Context, cluster capiv1alpha3.Cluster) error {
+	var err error
+
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cluster),
+	}
+	mpList := new(expcapiv1alpha3.MachinePoolList)
+	err = r.ctrlClient.List(ctx, mpList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, machinePool := range mpList.Items {
+		r.logger.LogCtx(ctx, "message", fmt.Sprintf("Ensuring %s label and 'ownerReference' fields on MachinePool '%s/%s'", capiv1alpha3.ClusterLabelName, machinePool.Namespace, machinePool.Name))
+
+		if machinePool.Labels == nil {
+			machinePool.Labels = make(map[string]string)
+		}
+		machinePool.Labels[capiv1alpha3.ClusterLabelName] = cluster.Name
+
+		// Set Cluster as owner of MachinePool
+		err = controllerutil.SetControllerReference(&cluster, &machinePool, r.scheme)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = r.ctrlClient.Update(ctx, &machinePool)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "message", fmt.Sprintf("Ensured %s label and 'ownerReference' fields on MachinePool '%s/%s'", capiv1alpha3.ClusterLabelName, machinePool.Namespace, machinePool.Name))
+	}
+
 	return nil
 }
 
-// Name returns the resource name.
-func (r *Resource) Name() string {
-	return Name
+func (r *Resource) ensureAzureMachinePools(ctx context.Context, cluster capiv1alpha3.Cluster) error {
+	var err error
+
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cluster),
+	}
+	mpList := new(expcapzv1alpha3.AzureMachinePoolList)
+	err = r.ctrlClient.List(ctx, mpList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, azureMachinePool := range mpList.Items {
+		r.logger.LogCtx(ctx, "message", fmt.Sprintf("Ensuring %s label and 'ownerReference' fields on AzureMachinePool '%s/%s'", capiv1alpha3.ClusterLabelName, azureMachinePool.Namespace, azureMachinePool.Name))
+
+		if azureMachinePool.Labels == nil {
+			azureMachinePool.Labels = make(map[string]string)
+		}
+		azureMachinePool.Labels[capiv1alpha3.ClusterLabelName] = cluster.Name
+
+		// Set Cluster as owner of AzureMachinePool
+		err = controllerutil.SetControllerReference(&cluster, &azureMachinePool, r.scheme)
```

Comment:
From the comments of the `SetControllerReference` function

> SetControllerReference sets owner as a Controller OwnerReference on controlled.
This is used for garbage collection of the controlled object and for
reconciling the owner object on changes to controlled (with a Watch + EnqueueRequestForOwner).
Since only one OwnerReference can be a controller, it returns an error if
there is another OwnerReference with Controller flag set.

IIUC this will fail because the `machinepoolownerreference` handler already sets the `MachinePool` as owner of `AzureMachinePool`, and `AzureMachinePool` can't contain two `OwnerReference` that are controllers.

Instead we should use the other method to add owner references without being controllers, like we do to set `Cluster` as owner of `MachinePool`:

```
	machinePool.OwnerReferences = capiutil.EnsureOwnerRef(machinePool.OwnerReferences, metav1.OwnerReference{
		APIVersion: capiv1alpha3.GroupVersion.String(),
		Kind:       "Cluster",
		Name:       cluster.Name,
		UID:        cluster.UID,
	})
```